### DISCUSSION TITLE
Integrate PosedPointcloud

### DISCUSSION THREAD

**Discussion by ceccocats** [ethz-asl/wavemap#63] [Q&A] [answered]
Hi,
I am trying to integrate pointclouds to a wavemap without using ros.
this is what i have done:
- modify code of PointcloudInputHandler to serialize cloud and position before every integration
- I launched `roslaunch wavemap_ros newer_college_os0_cloister.launch rosbag_dir:=/data/wavemap/ only_first_rosbag:=True`

Then I writed a simple app to integrate a PosedPointcloud from recorded data:
```
wavemap::PosedPointcloud<> readCloud()
{
    wavemap::PosedPointcloud<> pcloud;
    Eigen::Isometry3f ptf;
    deserializeMatrix(ptf.matrix());
    deserializeMatrix(pcloud.data());

    wavemap::Transformation3D tf(ptf.translation(), tk::math::Quaternionf(ptf.rotation()));
    pcloud.setPose(tf);
    std::cout << pcloud.getPose() << "\n";
    return pcloud;
}

main() {
    wavemap::HashedWaveletOctreeConfig config;
    wavemap::HashedWaveletOctree::Ptr wmap = std::make_shared<wavemap::HashedWaveletOctree>(config);
    // wavemap::io::fileToMap(mapFile, wmap);

    // integrate pointclouds
    wavemap::RayTracingIntegratorConfig integratorConfig;
    wavemap::RayTracingIntegrator::Ptr integrator =
      std::make_shared<wavemap::RayTracingIntegrator>(integratorConfig, wmap);

    for (int i = 0; i < 200; i++) {
        wavemap::PosedPointcloud<> wcloud = readCloud();
        std::cout << i << " integrate: " << wcloud.size() << "\n";
        integrator->integratePointcloud(wcloud);
    }
    wavemap::io::mapToFile(*wmap, "out.wmap");
    std::cout << "Map size: " << wmap->size() << "\n";
}
```

The issue is that I dont obtain the same map of the example but a totally wrong map.
Is something I am missing in this code? Or I should I check better the data input?

**Comment by victorreijgwart** [comment, marked as answer]
Hi @ceccocats,
The RayTracingIntegrator does not work well with wavelet-encoded maps (e.g. HashedWaveletOctree). I'd recommend using the `HashedWaveletIntegrator`. There's an example of how to use this integrator without ROS in [this reply](https://github.com/ethz-asl/wavemap/discussions/53#discussioncomment-7888171).
If you're interested in using wavemap without ROS, you might be happy to hear we're actively working on supporting this use case. We just finished porting the wavemap library into a pure CMake project (making usage of catkin optional) and are now starting to work on "pipelines", which will make it much easier to create and configure sensor_input+integrator->map streams. Given our current workload, it might still take several weeks until this is ready, but it will be something you could use in the not-too-distant future.
In the meantime, don't hesitate if you have any questions about manually interfacing maps and integrators manually.

**Reply by ceccocats** [reply]
WavemapServer in the sample does not use  HashedWaveletOctree with RayTracingIntegrator? why I cannot obtain the same result?

**Reply by victorreijgwart** [reply]
In the demos using the Newer College dataset, the default [Ouster OS0 configuration file](https://github.com/ethz-asl/wavemap/blob/main/ros/wavemap_ros/config/wavemap_ouster_os0.yaml) is used. These settings combine the HashedChunkedWaveletOctree map data structure with the HashedChunkedWaveletIntegrator. It will give equivalent results to the HashedWaveletOctree and HashedWaveletIntegrator combination, while being slightly faster and using less memory. The downside of the 'chunked' version is that its code is more complicated since it's more heavily optimized. So which combination is better depends on what you want to do with it (i.e. only build maps vs also working on the code's internals).

**Reply by ceccocats** [reply]
Ok I am missing a lot of stuff. As far as I understand I need a projector to convert pointcloud to depth image. Is it possible to use the pointcloud without projecting to an image?

Thanks

**Reply by victorreijgwart** [reply]
The integrator you're trying to use is projective - instead of iterating over the measurements and ray tracing, it iterates over the map and projects its voxels into the range image to compute their measurement update. So the range image is not optional. You can find the full motivation and explanation for the integrator in our paper, but the main advantages are that it's much more efficient and allows us to use a measurement model that's more accurate.
Is there a specific reason why you want to avoid using the projector and range image?
